src/server/io_utils.cc

```diff
@@ -0,0 +1,71 @@
+// Copyright 2022, DragonflyDB authors.  All rights reserved.
+// See LICENSE for licensing terms.
+//
+
+#include "server/io_utils.h"
+
+#include "base/flags.h"
+#include "base/logging.h"
+#include "server/error.h"
+
+using namespace std;
+
+namespace dfly {
+
+io::Result<size_t> BufferedStreamerBase::WriteSome(const iovec* vec, uint32_t len) {
+  // Write data to producer buffer.
+  buffered_++;
+  return io::BufSink{&producer_buf_}.WriteSome(vec, len);
+}
+
+void BufferedStreamerBase::NotifyWritten() {
+  // Wake up the consumer.
+  waker_.notify();
+  // Block if we're stalled because the consumer is not keeping up.
+  waker_.await([this]() { return !IsStalled() || IsStopped(); });
+}
+
+error_code BufferedStreamerBase::ConsumeIntoSink(io::Sink* dest) {
+  while (!IsStopped()) {
+    // Wait for more data or stop signal.
+    waker_.await([this]() { return buffered_ > 0 || IsStopped(); });
+
+    // Break immediately on cancellation.
+    if (cll_->IsCancelled()) {
+      waker_.notifyAll();  // Wake consumer if it missed it.
+      break;
+    }
+
+    // Swap producer and consumer buffers
+    std::swap(producer_buf_, consumer_buf_);
+    buffered_ = 0;
+
+    // If producer stalled, notify we consumed data and it can unblock.
+    waker_.notify();
+    RETURN_ON_ERR(dest->Write(consumer_buf_.InputBuffer()));
+
+    if (IsStopped()) {
```

Comment:
instead of checking here, you can unconditionally call `waker_.notifyAll();` after the loop.